src/federation/observer.rs

```diff
@@ -599,18 +755,163 @@ impl FederationObserver {
-                    &(item_index as i32),
-                    &(peer_id.to_usize() as i32),
-                    &(*height_vote as i32),
-                ],
-            )
-            .await?;
+        match wallet_ci {
+            WalletConsensusItem::BlockCount(height_vote) => {
+                dbtx.execute(
+                    "INSERT INTO block_height_votes VALUES ($1, $2, $3, $4, $5) ON CONFLICT DO NOTHING",
+                    &[
+                        &federation_id.consensus_encode_to_vec(),
+                        &(session_index as i32),
+                        &(item_index as i32),
+                        &(peer_id.to_usize() as i32),
+                        &(*height_vote as i32),
+                    ],
+                )
+                .await?;
+            }
+            WalletConsensusItem::PegOutSignature(peg_out_sig) => {
+                let peg_out_txid = peg_out_sig.txid.to_string();
+                let peg_out_txid_encoded =
+                    fedimint_core::TransactionId::from_str(peg_out_txid.as_str())
+                        .expect("Invalid on chain txid")
+                        .consensus_encode_to_vec();
+
+                dbtx.execute(
+                    "INSERT INTO wallet_withdrawal_transactions VALUES ($1, $2) ON CONFLICT DO NOTHING",
+                    &[
+                        &peg_out_txid_encoded,
+                        &federation_id.consensus_encode_to_vec(),
+                    ],
+                )
+                .await?;
+
+                dbtx.execute(
+                    "INSERT INTO wallet_withdrawal_signatures VALUES ($1, $2, $3, $4) ON CONFLICT DO NOTHING",
+                    &[
+                        &peg_out_txid_encoded,
+                        &(session_index as i32),
+                        &(item_index as i32),
+                        &(peer_id.to_usize() as i32),
+                    ],
+                )
+                .await?;
+
+                let num_sigs = dbtx
+                    .query_one(
+                        "
+                        SELECT COUNT(peer_id)::INT num_sigs
+                        FROM wallet_withdrawal_signatures
+                        WHERE on_chain_txid = $1
+                        GROUP BY on_chain_txid
+                        ",
+                        &[&peg_out_txid_encoded],
+                    )
+                    .await?
+                    .get::<_, i32>("num_sigs") as usize;
+
+                // 3n + 1 <= num_peers
+                // n <= (num_peers - 1) / 3
+                // threshold = num_peers - floor((num_peers - 1) / 3)
+                let threshold = {
+                    let num_peers = config.global.api_endpoints.len();
+                    num_peers - (num_peers - 1) / 3
+                };
+
+                if num_sigs < threshold {
+                    return Ok(());
+                }
+
+                // at this point, the transaction reached threshold and should broadcast
```

Comment:
This was sufficient to avoid the complicated retry logic in the esplora client for peg out transactions that didn't reach a threshold. If we have a threshold and don't see the tx on a block explorer yet, we should retry until we do.